packages/flutter_tools/lib/src/android/gradle_utils.dart

```diff
@@ -585,6 +517,104 @@ bool validateJavaGradle(Logger logger,
   return false;
 }
 
+// Returns compatibility information for the valid range of Gradle versions for
+// the specified Java version.
+JavaGradleCompat? getValidGradleVersionRangeForJavaVersion(
+  Logger logger, {
+  required String? javaV,
+}) {
+  if (javaV == null) {
+    logger.printTrace('Java version unknown ($javaV).');
+    return null;
+  }
+  for (final JavaGradleCompat data in _javaGradleCompatList) {
+    if (isWithinVersionRange(javaV, min: data.javaMin, max: data.javaMax, inclusiveMax: false)) {
+      return data;
+    }
+  }
+
+  logger.printTrace('Unable to determine valid Gradle version range for specified Java version.');
+  return null;
+}
+
+// Validate that the specified Java and Android Gradle Plugin (AGP) versions are
+// compatible with each other.
+//
+// Source of truth are the AGP release notes:
+// https://developer.android.com/build/releases/gradle-plugin
+bool validateJavaAgp(Logger logger,
+    {required String? javaV, required String? agpV}) {
+  if (javaV == null || agpV == null) {
+    logger.printTrace(
+        'Java version or AGP version unknown ($javaV, $agpV).');
+    return false;
+  }
+
+  // Check if AGP version is too old to perform validation.
+  if (isWithinVersionRange(agpV,
+      min: '1.0', max: oldestDocumentedJavaAgpCompatibilityVersion, inclusiveMax: false)) {
+    logger.printTrace('AGP Version: $agpV is too old to determine Java compatibility.');
+    return false;
+  }
+
+  if (isWithinVersionRange(agpV,
+        min: maxKnownAgpVersion, max: '100.100')) {
```

Comment:
I updated this to be `maxKnownAndSupportedAgpVersion` and made sure to update its description to mention that support means we are aware of its Java <-> AGP compatibility, since realistically, that's what our support means for that version in terms of how `maxKnownAndSupportedAgpVersion` is used.